syntax = "proto3";

package tiny_muduo_rpc;

enum MessageType
{
  REQUEST = 0;
  RESPONSE = 1;
  ERROR = 2; // not used
}

enum ErrorCode
{
  NO_ERROR = 0;
  WRONG_PROTO = 1;
  NO_SERVICE = 2;
  NO_METHOD = 3;
  INVALID_REQUEST = 4;
  INVALID_RESPONSE = 5;
  TIMEOUT = 6;
}

message RpcMessage
{
   MessageType type = 1;
   fixed64 id = 2;

  oneof optional_service{ string service = 3;}
  oneof optional_method{ string method = 4;}
  oneof optional_request{ bytes request = 5;}

  oneof optional_response{ bytes response = 6;}

  oneof optional_error{ ErrorCode error = 7;}
}
